backupccl: backup checkpoint files are now versioned and never deleted

DarrylWong opened the pull request:
This change teaches backup and restore to no longer overwrite or delete any checkpoint files,
and instead write a new version side by side with the old, and to read these new versioned
checkpoints. To make sure that older backup chains and mixed clusters are not broken by this change,
we only write to the progress directory if it is not a pure v22.1 or later cluster. This
logic can be removed in a post 22.1 release when we no longer write to the base directory.
This change is so we can in the future, no longer require write access for backups.

Addresses https://github.com/cockroachdb/cockroach/issues/76438

Release Justification: release blocker

Release Note enterprise: Checkpoint files are no longer overwritten and now versioned
and written side by side in the /progress directory. Temporary checkpoint files are
no longer written.

Jira issue: CRDB-14712
Epic CRDB-7586

Commit:
backupccl: backup checkpoint files in /progress are now versioned and never deleted

This change teaches backup and restore to no longer overwrite or delete any checkpoint files,
and instead write a new version side by side with the old, and to read these new versioned
checkpoints. To make sure that older backup chains and mixed clusters are not broken by this change,
we only write to the progress directory if it is not a pure v22.1 or later cluster. This
logic can be removed in a post 22.1 release when we no longer write to the base directory.
This change is so we can in the future, no longer require write access for backups.

Release Justification: release blocker

Release note (enterprise): Checkpoint files are no longer overwritten and now versioned
and written side by side in the /progress directory. Temporary checkpoint files are
no longer written.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76869)
<!-- Reviewable:end -->

adityamaru reviewed (changes requested):
flushing a round of comments, nice work on this overall!

DarrylWong commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1234,8 +1235,10 @@ func checkForPreviousBackup(
+// with how normal checkpoint files are stored and read.
 func tempCheckpointFileNameForJob(jobID jobspb.JobID) string {
-	return fmt.Sprintf("%s-%d", backupManifestCheckpointName, jobID)
+	return fmt.Sprintf("TEMP-%s-%d", backupManifestCheckpointName, jobID)
```
Hmm okay, just trying to wrap my head around this. I think I mostly agree (just trying to convince myself) with your logic.

I understand why the non-mixed version cluster would violate write-once semantics, and the fix seems pretty straight forward. Since it's 22.1 only we can just write a normal checkpoint right? 
Or more specifically, instead of calling `writeBackupManifestCheckpoint`, we just call `writeBackupManifest` and pass it in NewTimestampedCheckpointFileName() as the filename. Which would functionally write the same thing just no TEMP prefix. 
And then readManifestOnResume would just treat it as though there was already a checkpoint written which shouldn't be a problem? Since it doesn't really care if it was a temp checkpoint or a regular one? If that logic follows then I agree that it would work in this case.

Planning on 21.2 node logic seems sound.

Planning on a 22.1 node seems pretty similar to the logic for non-mixed to me. We don't actually care if it a temp checkpoint or actual checkpoint, we just want the latest one so it's okay to do away with the TEMP prefix. 
I guess the only thing I'm a little iffy about is the `non-timestamp suffixed since we are in a mixed version state` part. I understand why you'd have to do this for the base directory, cause that's where 21.2 nodes will look. But does this apply to the progress directory? 
Because otherwise this would mess with listing (since BACKUP-CHECKPOINT will always be at the top). I don't see why we couldn't version the checkpoint we write into the progress directory unless I'm missing something.

Assuming that my reasoning falls in line with yours, I agree that we can do away with the TEMP prefix + jobID suffix all together.

adityamaru replied:
>Since it doesn't really care if it was a temp checkpoint or a regular one?

exactly, yup. My slight preference here would be to have `writeBackupManifestCheckpoint` do all the timestamping and writing of the checkpoint and checksum, and `writeBackupManifest` only be responsible for writing the manifest. The special casing to check if the filename is a checkpoint or manifest makes me uneasy, so even if it means some duplication of code I think I'm okay with that.

> But does this apply to the progress directory?

You're right, so you're saying if a 22.1 node does the planning, and we're in a mixed version state, we should write a non-ts manifest checkpoint to the base directory and this way 21.2 and 22.1 nodes that may execute the backup job will see this?

DarrylWong replied:
> writeBackupManifestCheckpoint do all the timestamping and writing of the checkpoint and checksum,

Sure, good with me. I also agree the checking of filename to see what to do with it seems bad.

> we should write a non-ts manifest checkpoint to the base directory

Yeah so I was thinking if it's a mixed version, we just call the `writeNewCheckpointFile` WITH the timestamp suffixed. Which should do two things, write a non-ts checkpoint in the base directory, write a ts checkpoint in the progress directory.

1. If a 22.1 node executes the job, it will `readManifestOnResume` which first checks the progress directory, which will find our ts checkpoint properly since it looks in progress first.

2. If a 21.2 node executes the job, it will also call `readManifestOnResume`, but instead only check the base directory, which will find our non-ts checkpoint that we wrote.

Like I guess technically the write into the progress directory is unneeded, but as the logic stands right now, there is no further implementation needed to write to both. Whereas only writing to base would require some sort of flag or separate function to tell it it's not actually a regular checkpoint but rather a temp checkpoint. And I feel like that would be making the code a lot messier? Unless theres a reason I'm missing as to why we can't write to progress.

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1259,18 +1245,54 @@ func readLatestCheckpointFile(
 
-// writeNewCheckpointFile writes a new BACKUP-CHECKPOINT file to both the base
-// directory and latest-history directory, depending on cluster version.
+// writeNewCheckpointFile writes a new BACKUP-CHECKPOINT file to either
```
nit: writeNewCheckpointFile writes a new checkpoint (or checksum) file to the appropriate directory depending on the cluster version.

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1224,6 +1208,8 @@ func checkForPreviousBackup(
 }
 
 // tempCheckpointFileNameForJob returns temporary filename for backup manifest checkpoint.
+// It writes to its own temp-progress directory so to not interfere
```
nit: can we delete this comment?

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1259,18 +1245,54 @@ func readLatestCheckpointFile(
+	// version of the filename in the progress directory.
+	if exportStore.Conf().Provider == roachpb.ExternalStorageProvider_http {
+		r, err = exportStore.ReadFile(ctx, backupProgressDirectory+"/"+filename)
+		if err == nil {
```
hmm is this a bug? If err != nil we want to return the error here.

DarrylWong replied:
So basically if it's http, we can't list, so it just tries to read non ts version from progress directory. Otherwise if it isn't http, then we try to read the ts version using list.

If the http fails, it could be because we didn't find the file/directory, so we need to retry in the base directory. However, if there wasn't an error, then we don't need to retry, just return the reader.

If the non http list fails, it's because of a non file/directory not found error. Since list won't fail if it doesn't find anything. So thats why we don't have to retry the base directory in this case and can just return the error.

I added a comment to try and make it more clear, and am returning nil now to make it more obvious I meant to return that.

DarrylWong commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -413,8 +413,14 @@ func (b *backupResumer) Resume(ctx context.Context, execCtx interface{}) error {
 			}
 		}
 
+		defaultStore, err := p.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx, backupDetails.URI, p.User())
```
Yeah I agree it looks awkward but the problem I ran into is that writeManifestCheckpoint can be called from multiple places,  including functions like backup() which already has a ExternalStorage. Would it not be problematic to create a second ExternalStorage when one already exists?

adityamaru replied:
as long as you close the ExternalStorage after it is used, there should be no issues.

adityamaru commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10125,3 +10144,82 @@ func TestBackupRestoreSeparateExplicitIsDefault(t *testing.T) {
+// TestBackupCheckpointInBaseDirectory tests to see that a checkpoint
+// file in the base directory can be properly read when one is not found
+// in progress. This can occur when an older version node creates the backup.
+func TestBackupCheckpointInBaseDirectory(t *testing.T) {
```
Hmm I'm a little confused about what this test is trying to achieve. I get that an older version backup should write to the base directory, but an incremental backup does not really read the checkpoint file, right? It just operates on the previous backups manifest.

What I think we really want to test is a mixed version cluster running a backup, i.e. the node planning the backup is 21.2 but node executing is 22.1, and then node planning is 21.2 and executing is 22.1. This is much better suited for a roachtest like `backup/mixed-version` and I'd love to see that as a fast follow to this (and the LATEST) PR, wdyt?

DarrylWong replied:
Oh duh, that sounds right that it doesn't even look at the checkpoints. I wrote something similar for latest files (which should work with this approach?) and just hastily copied it over assuming it would work here too. I think a roachtest would work best for this I agree. And although I don't think we necessarily need one for LATEST the logic will all be there so why not?

adityamaru commented on pkg/cloud/userfile/filetable/file_table_read_writer.go:
```diff
@@ -926,17 +934,28 @@ func (f *FileToTableSystem) NewFileWriter(
-	//
-	// NB: userfile upload will error out on the client side if a file with the
-	// same name already exists.
-	err = f.deleteFileWithoutTxn(ctx, filename, e.ie)
```
hmm this is interesting, I guess we don't want to do this anymore in a non-mixed version cluster? Since backups become write only we should never encounter a constraint violation. I wonder if removing this introduces a regression to other writers to userfile. Apart from backup I think it's only EXPORT, and I can't see why it would need to overwrite files.

DarrylWong replied:
I think so but we can't get rid of it just yet right? Because of the latest file? After that one then yeah I don't see why we'd have to do it in non-mixed

adityamaru replied:
You're right, lets's add a TODO because until then userfile won't truly be write-only. Not a huge deal but worth fixing.

adityamaru commented on pkg/cloud/userfile/filetable/file_table_read_writer.go:
```diff
@@ -926,17 +934,28 @@ func (f *FileToTableSystem) NewFileWriter(
+	// For testing purposes we sometimes want to check to see that the files
+	// we write are not being overwritten, and skipping this delete is an
+	// easy way to do so as it will throw an error if the file already exists.
+	if !f.testingKnobs.EnforceNoOverwrite {
```
Can we get rid of this testing knob if we implement the new test approach?

adityamaru commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10125,3 +10144,82 @@ func TestBackupRestoreSeparateExplicitIsDefault(t *testing.T) {
 		sqlDB.Exec(t, "DROP DATABASE inc_fkdb;")
 	}
 }
+
```
general test comment: can we come up with a minimal unit test for the method that generates timestamped files. Maybe a subtest for each provider that writes a fixed number of timestamped files, lists them with limit 1, and then ensures that our lexicographic sorting algorithm is sound?

DarrylWong replied:
Didn't get a chance on friday to make buckets for gcs and azure and actually test it for those two(but definitely will), but the general test is there.

adityamaru reviewed (changes requested):
Thanks for iterating on this. Code changes look good to me, left comments about the tests now.

adityamaru commented:
Can we clean up the commit message/PR message to represent the current state of this PR. Also, the heading should be of the form:

`backupccl:...`

adityamaru commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10125,3 +10144,192 @@ func TestBackupRestoreSeparateExplicitIsDefault(t *testing.T) {
+			require.NoError(t, err)
+			for _, checkpoint := range checkpoints {
+				var desc []byte
+				require.NoError(t, writeNewCheckpointFile(ctx, execCfg.Settings, store, checkpoint, desc))
```
one quick comment, can we rand.Shuffle the `checkpoints` before we write them? This way we're doubly sure the List with limit 1 is independent of the order of writing.

DarrylWong commented:
Not sure, the tests pass locally for me. It looked like it passed last night though, and all I changed was I accidentally forgot to return an error and went back to returning it

https://github.com/cockroachdb/cockroach/pull/76869/files#diff-69f3e78031edfb928564a413841d2fa8c580594390c272cadb3e37fdcb01deb4R1098-R1100

And the rand.shuffle stuff which i don't see how that would affect other test cases.

DarrylWong commented:
So basically the bug was really silly, just I forgot to account for the fact that `readLatestCheckpointFile` is also used to read in checksum files too, and it just returned the actual checkpoint  This would cause it to just compare the checksum of checkpoint manifest with itself, which i imagine would just always be true.

To account for checksums though, I need an if else in three different locations to see if I need the checksum suffix or not. It works but I'm wondering if we just want to make a `readLatestCheckpointChecksumFile` in order to get rid of the `strings.HasSuffix()` check.

https://github.com/cockroachdb/cockroach/blob/f13e8edf02e383e338693461d4b2aa6a02c4763b/pkg/ccl/backupccl/manifest_handling.go#L1316-L1319

Interestingly enough, fixing this bug caused another bug to start causing problems. Turns out I wasn't properly accounting for incremental backups when deciding to call read manifest or read checkpoint manifest. Since incremental backup manifests have the directory before it, I was accidentally sending it to `readLatestCheckpointFile` which would try to list, fail, and then read the filename from the base directory, which would work, covering up the bug. Fixed by checking if it was suffixed instead of equal.

https://github.com/cockroachdb/cockroach/blob/f13e8edf02e383e338693461d4b2aa6a02c4763b/pkg/ccl/backupccl/manifest_handling.go#L248

I also added the t=0 checkpoint stuff.  The problem I ran into is that the very first one is written in Resume, but resume can be called more than once. So no matter what we have to check to see if there exists a checkpoint already I think. Maybe there is a better way to do it but this is what I could think of.

https://github.com/cockroachdb/cockroach/blob/f13e8edf02e383e338693461d4b2aa6a02c4763b/pkg/ccl/backupccl/manifest_handling.go#L1325-L1327

adityamaru commented:
Hmmm I think I'm a little more confused than I was before the latest set of changes. Scheduling some time offline to discuss this PR since I think that'll be easier.

DarrylWong commented:
In regard to `strings.HasSuffix(filename, backupManifestChecksumSuffix) ` part, its down to just one usage since the sentinel value for http will just be the same as filename. I think in that case we can probably get away with not splitting up the functions? Although it looking at the code splitting it should be trivial (hopefully).

adityamaru commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
+	}
+
+	// Write the checksum file after we've successfully wrote the checkpoint.
+	checksum, err := getChecksum(descBuf)
```
Move this above the HTTP check and then do the following:

```
if !execCfg.Settings.Version.IsActive(ctx, clusterversion.BackupDoesNotOverwriteLatestAndCheckpoint) {
		// We want to overwrite the latest checkpoint in the base directory,
		// just write to the non versioned BACKUP-CHECKPOINT file.
		err = cloud.WriteFile(ctx, defaultStore, backupManifestCheckpointName, bytes.NewReader(descBuf))
if err != nil {
    return err
}

// Write checksum.
return cloud.WriteFile(ctx, defaultStore, backupManifestCheckpointName+backupManifestChecksumSuffix, bytes.NewReader(checksum))
	}


// Move HTTP check and all timestamping logic here.
```

adityamaru commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
+
+	// We timestamp the checkpoint files in order to enforce write once
+	// backups. By timestamping with monotonically increasing timestamps
+	// we avoid ever overwriting a checkpoint file.
```
nit: add a little comment around how we will consume these timestamped checkpoints

When the job goes to read these timestamped files, it will List the checkpoints and pick the file whose name is lexicographically sorted to the top. This will be the last checkpoint we write, for details refer to `newTimestampedCheckpointFileName`.

adityamaru commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
+	// we avoid ever overwriting a checkpoint file.
+	filename := newTimestampedCheckpointFileName()
+
+	// HTTP storage does not support listing, and so the very first time
```
The comment should give some more context to a new user.

HTTP storage does not support listing and so we cannot rely on the above-mentioned List method to return us the latest checkpoint file. Instead, we will write a checkpoint once with a well-known filename, and teach the job to always reach for that filename in the face of a resume. We may lose progress, but this is a cost we are willing to pay to uphold write-once semantics.

adityamaru commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
+	// that we can always find without listing.
+	if defaultStore.Conf().Provider == roachpb.ExternalStorageProvider_http {
+		if r, err := defaultStore.ReadFile(ctx, backupProgressDirectory+"/"+backupManifestCheckpointName); err != nil {
+			filename = backupManifestCheckpointName
```
I don't like that we are doing this unconditionally for all errors. Can you add a TODO to catch specific file not found or directory not found errors here?

Also a comment saying, since we did not find the checkpoint file this is the first time we are going to write a checkpoint, so write it with the well known filename.

adityamaru commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
+		if r, err := defaultStore.ReadFile(ctx, backupProgressDirectory+"/"+backupManifestCheckpointName); err != nil {
+			filename = backupManifestCheckpointName
+		} else {
+			r.Close(ctx)
```
we should handle the error

adityamaru commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -9888,3 +9907,200 @@ func TestRestoreOnFailOrCancelAfterPause(t *testing.T) {
+		// Occasionally, we call newTimestampedCheckpointFileName() in succession
+		// too fast and the timestamp is the same. So wait for a moment to
+		// avoid that.
+		time.Sleep(time.Millisecond)
```
loll

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -223,52 +223,89 @@ func decompressData(ctx context.Context, mem *mon.BoundAccount, descBytes []byte
+	manifestReader ioctx.ReadCloserCtx,
+	checksumReader ioctx.ReadCloserCtx,
+) (BackupManifest, int64, error) {
+	defer manifestReader.Close(ctx)
```
lets move this defer to after where the reader is initialized. This method is just taking the reader as a param, it shouldn't be responsible for closing it.

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -223,52 +223,89 @@ func decompressData(ctx context.Context, mem *mon.BoundAccount, descBytes []byte
 		// If there is a checksum file present, check that it matches.
-		defer checksumFile.Close(ctx)
-		checksumFileData, err := ioctx.ReadAll(ctx, checksumFile)
+		defer checksumReader.Close(ctx)
```
ditto as above

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1259,38 +1274,60 @@ func readLatestCheckpointFile(
+	// be at the top, so just grab the first filename.
+	err = exportStore.List(ctx, backupProgressDirectory, "", func(p string) error {
+		p = strings.TrimPrefix(p, "/")
+		checkpoint = strings.TrimSuffix(p, backupManifestChecksumSuffix)
```
nit: comment here that the first file returned by List could be the checkpoint or its checksum but we're only concerned with its timestamped prefix. We resolve the checksum and checkpoint separately below.

adityamaru commented on pkg/ccl/backupccl/manifest_handling.go:
```diff
@@ -1259,38 +1274,60 @@ func readLatestCheckpointFile(
+		return nil
+	}, 1 /*limit*/)
+	// If the list failed because the storage used does not support listing,
+	// such as http we can try reading the non ts backup checkpoint
```
nit: s/ts/timestamped/g

adityamaru reviewed (approved):
Thanks for the cleanup looks a lot better. I would love some tests for http and its special casing, and mixed version. I know you've got both those in your head too but if you want to create an issue to capture that that'd be cool too. Lets get this merged so we have some bake time.

DarrylWong commented:
bors r+

adityamaru commented:
blathers backport 22.1

dt commented:
There are a lot of collapsed comments above so maybe this was covered, but could we just use the List() API as-is and have our listing function return a non-nil error after it gets its first result? e.g. io.EOF a new cloud.ErrListingDone or something?

DarrylWong commented:
Hmm, I think that would work. I can make a branch and quickly see if that breaks anything. 

Is that preferable so we don't have to change the List() API?

dt commented:
I was just thinking it'd reduce the size of the backport, since we wouldn't need to touch the cloud API -- just return an error from our listing fn once we've seen the first result, watch that error pop back out of the list call and then discard it as expected, with no added implementation complexity / room for bugs or differences between implementations.

That said, I suppose adding limit param to the API could allow some implementations to tune their page sizes of their underlying RPCs which could avoid some wasted work?

I think I'd slightly lean towards no pkg/cloud change just to shrink the backporting diff but could go either way.

DarrylWong commented:
Makes sense to me, went ahead and did it real quick.

https://github.com/cockroachdb/cockroach/pull/79359

Doesn't seem to have broken anything and was a pretty simple change. Plus the limit was only used in one spot anyway and that seems to work. I'll let it run on CI though and maybe see what Aditya has to say.
